File: tests/selenium/src/test/java/Selenium/Utilities/Holidays.java
/*
 * Copyright (C) 2024 Mandelkow
 *
 * Dienstplan Apotheke
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package Selenium.Utilities;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.Month;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;

public class Holidays {

    private Map<LocalDate, String> holidays;
    private int yearOnConstruct;

    public Holidays(int year) {
        // Add holidays to the map
        yearOnConstruct = year;
        this.holidays = new HashMap<>();
        holidays.put(LocalDate.of(year, Month.JANUARY, 1), "Neujahr");
        holidays.put(LocalDate.of(year, Month.MAY, 1), "Tag der Arbeit");
        holidays.put(LocalDate.of(year, Month.OCTOBER, 3), "Tag der Deutschen Einheit");
        holidays.put(LocalDate.of(year, Month.OCTOBER, 31), "Reformationstag");
        holidays.put(LocalDate.of(year, Month.DECEMBER, 25), "1. Weihnachtsfeiertag");
        holidays.put(LocalDate.of(year, Month.DECEMBER, 26), "2. Weihnachtsfeiertag");
        holidays.put(LocalDate.of(year, Month.MARCH, 8), "Internationaler Frauentag");
        // Add more holidays as needed
        LocalDate easterDate = Easter.computeEasterDateWithConwayAlgorithm(year);
        holidays.put(easterDate.minusDays(2), "Karfreitag");
        holidays.put(easterDate, "Ostersonntag");
        holidays.put(easterDate.plusDays(1), "Ostermontag");
        holidays.put(easterDate.plusDays(39), "Himmelfahrt");
        holidays.put(easterDate.plusDays(49), "Pfingstsonntag");
        holidays.put(easterDate.plusDays(50), "Pfingstmontag");
    }

    public boolean isHoliday(LocalDate date) throws Exception {
        // Check if the given date is a holiday
        if (yearOnConstruct != date.getYear()) {
            throw new Exception("The year of the date must be: " + yearOnConstruct + " but the date is " + date.format(DateTimeFormatter.ISO_DATE));
        }
        return holidays.containsKey(date);
    }

}

File: src/php/classes/PDR/DateTime/Holidays.php
<?php

/*
 * Copyright (C) 2025 Mandelkow
 *
 * Dienstplan Apotheke
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

namespace PDR\DateTime;

class Holidays {

    /**
     * @var array<int, array<string, \PDR\DateTime\Holiday>>
     */
    private $holidays = array();

    public function __construct() {

    }

    private function loadYear(int $year): void {
        if (isset($this->holidays[$year])) {
            return;
        }

        $configuration = new \PDR\Application\Configuration();

        $this->holidays[$year] = $this->getHolidays(
                $year,
                $configuration->getCountryCode(),
                $configuration->getStateCode()
        );
    }

    /**
     * Returns an associative array of holidays, keyed by date (Y-m-d).
     *
     * @param int $year An integer representing the holiday year.
     * @param string $country ISO 3166 ALPHA-2 country code (e.g. "DE").
     * @param string $state ISO 3166-2 state code (e.g. "DE-MV").
     * @return array<string, PDR\DateTime\Holiday>
     */
    private function getHolidays(int $year, string $country = "DE", string $state = "DE-MV"): array {
        switch ($country) {
            case "DE":
                $germanHolidaysProvider = new \PDR\DateTime\GermanHolidays($year, $state);
                $holidays = $germanHolidaysProvider->getListOfHolidays();
                break;
            case "FR":
                $frenchHolidaysProvider = new \PDR\DateTime\FrenchHolidays($year, $state);
                $holidays = $frenchHolidaysProvider->getListOfHolidays();
                break;
            case "EN":
                $britishHolidaysProvider = new \PDR\DateTime\BritishHolidays($year, $state);
                $holidays = $britishHolidaysProvider->getListOfHolidays();
                break;
            default:
                throw new \Exception("This country is not supported yet.");
        }
        return $holidays;
    }

    /**
     * Test if a day is a holiday.
     *
     * This function returns FALSE if a day is not a holiday.
     * This function returns TRUE on a holiday.
     * @param DateTime $dateObject
     *
     * @return boolean.
     */
    public function isHoliday(\DateTime $dateObject): bool {
        $year = (int) $dateObject->format('Y');

        $this->loadYear($year);

        return isset(
                $this->holidays[$year][$dateObject->format('Y-m-d')]
        );
    }

    public function getHolidayOnDate(\DateTime $dateObject): \PDR\DateTime\Holiday {
        $dateString = $dateObject->format('Y-m-d');
        $year = (int) $dateObject->format('Y');
        $this->loadYear($year);
        if (isset($this->holidays[$year][$dateString])) {
            return $this->holidays[$year][$dateString];
        }
        throw new \Exception("There is no holiday found on this date.");
    }
}
